sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureSasCredential.java

```diff
@@ -6,28 +6,50 @@
 import com.azure.core.util.logging.ClientLogger;
 
 import java.util.Objects;
+import java.util.function.Function;
 
 /**
  * Represents a credential that uses a shared access signature to authenticate to an Azure Service.
  */
 public final class AzureSasCredential {
     private final ClientLogger logger = new ClientLogger(AzureSasCredential.class);
     private volatile String signature;
+    private final Function<String, String> sanitationFunction;
 
     /**
      * Creates a credential that authorizes request with the given shared access signature.
+     * <p>
+     * The {@code signature} passed is assumed to be sanitized of any illegal characters. This constructor is
+     * effectively the same as calling {@link #AzureSasCredential(String, Function) new AzureSasCredential(signature,
+     * null))}.
      *
      * @param signature The shared access signature used to authorize requests.
      * @throws NullPointerException If {@code signature} is {@code null}.
      * @throws IllegalArgumentException If {@code signature} is an empty string.
      */
     public AzureSasCredential(String signature) {
+        this(signature, null);
+    }
+
+    /**
+     * Creates a credential that authorizes request within the given shared access signature.
+     * <p>
+     * If {@code sanitationFunction} is non-null the {@code signature}, and all {@link #update(String) updated
+     * signatures}, will be sanitized using the function.
+     *
+     * @param signature The shared access signature used to authorize requests.
+     * @param sanitationFunction An optional sanitation function to sanitize the {@code signature}.
+     * @throws NullPointerException If {@code signature} is {@code null}.
+     * @throws IllegalArgumentException If {@code signature} is an empty string.
+     */
+    public AzureSasCredential(String signature, Function<String, String> sanitationFunction) {
         Objects.requireNonNull(signature, "'signature' cannot be null.");
         if (signature.isEmpty()) {
             throw logger.logExceptionAsError(new IllegalArgumentException("'signature' cannot be empty."));
         }
 
-        this.signature = signature;
+        this.sanitationFunction = sanitationFunction;
+        this.signature = sanitizeSignature(signature, sanitationFunction);
```

Comment:
What happens if the signature is already sanitized? Should the sanitationFunction be idempotent? Also, should we call this `signatureEncoder` instead of `sanitation`?

Reply:
Renamed and I updated the comment slightly to indicate `signatureEncoder` should be as idempotent as possible to reduce the change of double encoding errors.